### ISSUE TITLE
Winsock clean-up atexit callback commented?

### ISSUE THREAD

**Issue by DoumanAsh** [alexcrichton/net2-rs#48]
Is there a reason that [these](https://github.com/rust-lang-nursery/net2-rs/blob/master/src/sys/windows/mod.rs#L44-L48) lines are commented out?
atexit() allows to register multiple callbacks so it shouldn't be harmful to register one more.

**Comment by alexcrichton** [comment]
In general MSVC doesn't have a great way to register atexit callbacks w/o pulling in the CRT, which isn't always desired. Put another way, I don't know of a great "register an at exit callback" location.

**Comment by DoumanAsh** [comment]
I'm sorry, i didn't exactly get what is the problem?
You don't know where is right place to invoke `atexit()` or there is some MSVC specifics in regard of atexit callback?

Rust itself is using atexit as far as i know, so i was wondering what could be an issue...

**Comment by alexcrichton** [comment]
I don't know of any place to put this, but PRs are always welcome!

**Comment by DoumanAsh** [comment]
@alexcrichton After looking into Rust implementation it seems i was mistaken and it is using own preudo `atexit`.
It seems to store at start callback in heap and runs it before shutting down.

Considering the way it is designed i assume there is possibility to bypass libc `atexit` callbacks.

If so i believe it would be better to have a different approach, considering following points:
1. Rust MUST initialize Winsock always as long as it is using it.
2. net2-rs relies on libstd and therefore it can rely on this initialization.

What do you think if we completely remove winsock initialization from crate?
I don't really like it, but it is better than come up with some own hacks (considering Rust lacking any API to have atexit, no idea how to work around it)

**Comment by alexcrichton** [comment]
Actually yeah that sounds pretty reasonable to me, basically the initialization here is to just create a socket from the standard library which would transitively initialize it.

**Comment by DoumanAsh** [comment]
@alexcrichton I made a quick test and it seems that i mistook Rust code. It turns out that Rust doesn't initialize winsock at a start. Instead Rust calls init on each net API function.

**UPD:** Since `net2` is bypassing initialize libstd `net` we need to consider that. I don't think it is good idea to initialize winsock inside `net2` but not sure how to work around it properly

**UPD 2:** I used `libc::atexit` and it seems that it is called properly. Rust uses winapi ExitProcess to exit so i had some doubts about wether it would invoke atexit callbacks, but it seems to invoke it.
Should then `net2` to use `libc::atexit()`?
